proposals/csharp-7.0/tuples.md

```diff
@@ -0,0 +1,437 @@
+
+```csharp
+object o = (a: 1, b: 2);           // boxing conversion
+var t = ((int moo, int boo))o;     // unboxing conversion
+```
+
+## Additions to [Variables](../../spec/variables.md)
+
+> The following text should be added at the end of the [Variables](../../spec/variables.md) section.
+
+### "Discards"
+
+The `_` can be used as a *discard* under these rules:
+
+- A standalone `_` is a discard when no `_` is defined in scope.
+- A "designator" `var _` or `T _` in deconstruction, pattern matching and out vars is a discard.
+- Discards are like unassigned variables, and do not have a value. They can only occur in contexts where they are assigned to.
+
+Examples:
+
+```csharp
+M(out _, out var _, out int _); // three out variable discards
+(_, var _, int _) = GetCoordinates(); // deconstruction into discards
+if (x is var _ && y is int _) { ... } // discards in patterns
+```
+
+## Changes to [Conversions](../spec/conversions.md)
+
+> The following text should be added to [Identity conversions](../../spec/conversions.md#identity-conversions), after the bullet point on `object` and `dynamic`:
+
+* Element names are immaterial to tuple conversions. Tuples with the same types in the same order are identity convertible to each other or to and from corresponding underlying `ValueTuple` types, regardless of the names.
+
+```csharp
+var t = (sum: 0, count: 1);
+
+System.ValueTuple<int, int> vt = t;  // identity conversion
+(int moo, int boo) t2 = vt;          // identity conversion
+
+t2.moo = 1;
+```
+
+> note:
+>
+> An element name at one position on one side of a conversion, and the same name at another position on the other side almost certainly have bug in the code:
+
+```csharp
+(string first, string last) GetNames() { ... }
+(string last, string first) names = GetNames(); // Oops!
+```
+
+> Compilers should issue a warning for the preceding code.
+>
+> endnote.
+
+### Boxing conversions
+
+> The following text should be added to [Boxing conversions](../../spec/conversions.md#boxing-conversions) after the first paragraph:
+
+Tuples, like all value types, have a boxing conversion. Importantly, the names aren't part of the runtime representation of tuples, but are tracked only by the compiler. Thus, once you've "cast away" the names, you cannot recover them. In alignment with the identity conversions, a boxed tuple will unbox to any tuple type that has the same element types in the same order.
+
+### Tuple conversions
+
+> This section should be added after [Implicit enumeration conversions](../../spec/conversions.md#implicit-enumeration-conversions)
+
+Tuple types and expressions support a variety of conversions by "lifting" conversions of the elements into overall *tuple conversion*.
+For the classification purpose, all element conversions are considered recursively. For example: To have an implicit conversion, all element expressions/types must have implicit conversions to the corresponding element types.
```

Comment:
>considered recursively [](start = 60, length = 22)

Then they need to be specified recursively.

Reply:
Rex has addressed this in the feature branch.